python/lsst/daf/butler/column_spec.py

```diff
@@ -88,6 +89,102 @@
+    """Class that knows how to serialize and deserialize column values."""
+
+    @abstractmethod
+    def serialize(self, value: Any) -> Any:
+        """Convert column value to something that can be serialized.
+
+        Parameters
+        ----------
+        value : `Any`
+            Column value to be serialized.
+
+        Returns
+        -------
+        value : `Any`
+            Column value in serializable format.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def deserialize(self, value: Any) -> Any:
+        """Convert serialized value to column value.
+
+        Parameters
+        ----------
+        value : `Any`
+            Serialized column value.
+
+        Returns
+        -------
+        value : `Any`
+            Deserialized column value.
+        """
+        raise NotImplementedError
+
+
+class _DefaultColumnValueSerializer(ColumnValueSerializer):
+    """Default implementation of serializer for basic types."""
+
+    def serialize(self, value: Any) -> Any:
+        # Docstring inherited.
+        return value
+
+    def deserialize(self, value: Any) -> Any:
+        # Docstring inherited.
+        return value
+
+
+class _TypeAdapterColumnValueSerializer(ColumnValueSerializer):
+    """Implementation of serializer that uses pydantic type adapter."""
+
+    def __init__(self, type_adapter: pydantic.TypeAdapter):
+        # Docstring inherited.
+        self._type_adapter = type_adapter
+
+    def serialize(self, value: Any) -> Any:
+        # Docstring inherited.
+        return value if value is None else self._type_adapter.dump_python(value)
+
+    def deserialize(self, value: Any) -> Any:
+        # Docstring inherited.
+        return value if value is None else self._type_adapter.validate_python(value)
+
+
+class _DateTimeColumnValueSerializer(ColumnValueSerializer):
+    """Implementation of serializer for ingest_time column. That column can be
+    either in native database time appearing as `datetime.datetime` on Python
+    side or integer nanoseconds appearing as astropy.time.Time. We use pydantic
+    type adapter for astropy time, which serializes it into integer
+    nanoseconds. datetime is converted to string representation to distinguish
+    it from integer nanoseconds (timezone handling depends entirely on what
+    database returns).
+    """
+
+    def __init__(self) -> None:
+        self._astropy_adapter = pydantic.TypeAdapter(SerializableTime)
+
+    def serialize(self, value: Any) -> Any:
+        # Docstring inherited.
+        if value is None:
+            return None
+        elif isinstance(value, datetime.datetime):
+            return value.isoformat()
+        else:
+            return self._astropy_adapter.dump_python(value)
+
+    def deserialize(self, value: Any) -> Any:
+        # Docstring inherited.
+        if value is None:
+            return None
+        elif isinstance(value, str):
+            return datetime.datetime.fromisoformat(value)
```

Comment:
I understand wanting to accept `datetime.datetime` and `str` from users and databases for backwards compatibility, but I don't understand why we'd want to return a `datetime.datetime` to users or serialize to `str` going forward.

My recollection is that we wanted to eventually switch to astropy/nsec for ingest_time everywhere, but it's a tricky migration due to the timezone stuff and we don't really have a sound plan for doing it.  But I'd like to get the in-memory Python types consistent.

If that's all true, I think we could definitely support coercion from `str` or `datetime.datetime` in `pydantic_utils._deserialize_time`, and that might be enough to replace the new serialization dispatch logic here.  But it seems likely I'm not thinking about at least one of the places where we need to do the time type conversions.

Reply:
I chose this for compatibility between direct butler and remote butler. Queries with direct butler return whatever sqlalchemy does for that column, which is `datetime` in current schema. It may be surprising for clients if remote butler returned astropy time instead for the same query. We could convert all `datetime` instances returned from queries to astropy, then serialization would need to handle only one type of time. 
These new methods are only used to (de)serialize query results, I'm not sure what happens when the query itself includes time instances in `bind`, need to check.

Reply:
Thus far it hasn't actually been possible to get the `ingest_date` back from a query - there's no interface in the Registry query system that can do it, and the "general" result type is the only way to do it in the new one.  So I think we've still got until we open up `butler._query*` as a public interface to get this this fixed, and IMO that means coercing to astropy.  I'm happy for that to be another ticket, though.

Reply:
OK, I'll make a new ticket for that.